server/server.go

```diff
@@ -416,13 +416,17 @@ func filterpath(peer *Peer, path, old *table.Path) *table.Path {
 		}
 
 		if ignore {
-			if !path.IsWithdraw && old != nil && old.GetSource().Address.String() != peer.ID() && old.GetSource().AS != peer.fsm.pConf.State.PeerAs {
-				// we advertise a route from ebgp,
-				// which is the old best. We got the
-				// new best from ibgp. We don't
-				// advertise the new best and need to
-				// withdraw the old.
-				return old.Clone(true)
+			if !path.IsWithdraw && old != nil {
+				oldSource := old.GetSource()
+				if oldSource.Address == nil || oldSource.Address.String() != peer.ID() && oldSource.AS != peer.AS() {
```

Comment:
I think that old.IsLocal() is better than "oldSource.Address == nil"

Reply:
Yes, I think `old.IsLocal()` is better to read code, but `old.IsLocal()` calls `old.GetSource()` internally and `GetSource()` can NOT be lightweight getter function, because `GetSource()` need to draw "root" instance of this path. So I wanted to reduce the calls of `GetSource()`.
If drawing root does not take cost so high, better to use `old.IsLocal()` as you said.

Reply:
I prefer the code readability until someone complains with the performance results showing that old.IsLocal() actually degrades the performance.

Reply:
I see, I will use `IsLocal()` instead.